File: macos/QuickProtectTests/PreserveEnrichmentFlagsTests.swift
import XCTest

/// `Camera.preservingEnrichmentFlags` carries PTZ capability flags forward
/// across Integration-API refreshes (whose payloads never include them), so a
/// panel reopen inside the enrichment throttle window doesn't wipe PTZ chips,
/// overlay, and movement. Mirrors the .NET PreserveEnrichmentFlagsTests.
final class PreserveEnrichmentFlagsTests: XCTestCase {

    private func camera(_ id: String, isPtz: Bool = false, canZoom: Bool = false) -> Camera {
        Camera(id: id, name: id, state: "CONNECTED", channels: [],
               isPtz: isPtz, canZoom: canZoom, secondaryLens: nil)
    }

    func testCarriesPtzFlagForwardById() {
        let previous = [camera("a", isPtz: true), camera("b")]
        let fresh = [camera("a"), camera("b")]

        let merged = Camera.preservingEnrichmentFlags(from: previous, into: fresh)
        XCTAssertTrue(merged[0].isPtz, "enriched flag must survive an unenriched refresh")
        XCTAssertFalse(merged[1].isPtz)
    }

    func testCarriesZoomFlagIndependentlyOfPtz() {
        // A zoom-only camera (optical lens, no pan/tilt motor) keeps just canZoom.
        let previous = [camera("z", canZoom: true)]
        let fresh = [camera("z")]

        let merged = Camera.preservingEnrichmentFlags(from: previous, into: fresh)
        XCTAssertTrue(merged[0].canZoom)
        XCTAssertFalse(merged[0].isPtz)
    }

    func testDoesNotFlagPreviouslyUnflaggedCameras() {
        let previous = [camera("a", isPtz: true, canZoom: true), camera("b")]
        let fresh = [camera("b")]

        let merged = Camera.preservingEnrichmentFlags(from: previous, into: fresh)
        XCTAssertFalse(merged[0].isPtz)
        XCTAssertFalse(merged[0].canZoom)
    }

    func testRemovedCameraIsNotResurrected() {
        let previous = [camera("gone", isPtz: true), camera("kept", isPtz: true)]
        let fresh = [camera("kept")]

        let merged = Camera.preservingEnrichmentFlags(from: previous, into: fresh)
        XCTAssertEqual(merged.map(\.id), ["kept"])
        XCTAssertTrue(merged[0].isPtz)
    }

    func testNewCameraStaysUnflagged() {
        let previous = [camera("a", isPtz: true)]
        let fresh = [camera("a"), camera("new")]

        let merged = Camera.preservingEnrichmentFlags(from: previous, into: fresh)
        XCTAssertTrue(merged[0].isPtz)
        XCTAssertFalse(merged[1].isPtz)
        XCTAssertFalse(merged[1].canZoom)
    }

    func testOnlyEverSetsFlagsTrue() {
        // A fresh camera already flagged true (e.g. from a cached list) must not
        // be cleared by a previous list where it was false — the merge only
        // sets flags, it never clears; clearing is the enrichment's job.
        let previous = [camera("a", isPtz: false, canZoom: false), camera("b", isPtz: true)]
        let fresh = [camera("a", isPtz: true, canZoom: true), camera("b")]

        let merged = Camera.preservingEnrichmentFlags(from: previous, into: fresh)
        XCTAssertTrue(merged[0].isPtz)
        XCTAssertTrue(merged[0].canZoom)
        XCTAssertTrue(merged[1].isPtz)
    }

    func testEmptyPreviousListLeavesFreshUntouched() {
        let fresh = [camera("a"), camera("b")]

        let merged = Camera.preservingEnrichmentFlags(from: [], into: fresh)
        XCTAssertEqual(merged.map(\.id), ["a", "b"])
        XCTAssertFalse(merged.contains { $0.isPtz || $0.canZoom })
    }

    func testEmptyFreshListStaysEmpty() {
        let previous = [camera("a", isPtz: true)]

        let merged = Camera.preservingEnrichmentFlags(from: previous, into: [])
        XCTAssertTrue(merged.isEmpty)
    }
}

File: dotnet/tests/QuickProtect.Core.Tests/PreserveEnrichmentFlagsTests.cs
using QuickProtect.Core.Models;
using QuickProtect.Core.Services;
using Xunit;

namespace QuickProtect.Core.Tests;

/// <summary>
/// Tests for the enrichment-flag carry-over on camera-list refresh: Integration
/// API responses have no featureFlags, so a fetch inside the PTZ-enrichment
/// throttle window must not lose the classic-API IsPtz/CanZoom flags (the
/// regression that hid PTZ overlays after a panel reopen).
/// </summary>
public class PreserveEnrichmentFlagsTests
{
    private static Camera Cam(string id, bool ptz = false, bool zoom = false)
        => new() { Id = id, Name = id, IsPtz = ptz, CanZoom = zoom };

    [Fact]
    public void Carries_flags_onto_fresh_list_by_id()
    {
        var previous = new[] { Cam("a", ptz: true, zoom: true), Cam("b") };
        var fresh = new[] { Cam("a"), Cam("b") };
        var result = ProtectService.PreserveEnrichmentFlags(previous, fresh);
        Assert.True(result[0].IsPtz);
        Assert.True(result[0].CanZoom);
        Assert.False(result[1].IsPtz);
        Assert.False(result[1].CanZoom);
    }

    [Fact]
    public void Never_clears_flags_already_on_the_fresh_list()
    {
        var previous = new[] { Cam("a") };
        var fresh = new[] { Cam("a", ptz: true, zoom: true) };
        var result = ProtectService.PreserveEnrichmentFlags(previous, fresh);
        Assert.True(result[0].IsPtz);
        Assert.True(result[0].CanZoom);
    }

    [Fact]
    public void Ignores_cameras_that_left_the_list()
    {
        var previous = new[] { Cam("gone", ptz: true) };
        var fresh = new[] { Cam("new") };
        var result = ProtectService.PreserveEnrichmentFlags(previous, fresh);
        Assert.False(result[0].IsPtz);
    }

    [Fact]
    public void Empty_previous_list_returns_fresh_unchanged()
    {
        var fresh = new[] { Cam("a", ptz: true) };
        var result = ProtectService.PreserveEnrichmentFlags(Array.Empty<Camera>(), fresh);
        Assert.Same(fresh, result);
        Assert.True(result[0].IsPtz);
    }

    [Fact]
    public void Zoom_and_ptz_carry_independently()
    {
        var previous = new[] { Cam("a", ptz: true), Cam("b", zoom: true) };
        var fresh = new[] { Cam("a"), Cam("b") };
        var result = ProtectService.PreserveEnrichmentFlags(previous, fresh);
        Assert.True(result[0].IsPtz);
        Assert.False(result[0].CanZoom);
        Assert.False(result[1].IsPtz);
        Assert.True(result[1].CanZoom);
    }
}
